src/D2L.CodeStyle.Analyzers/ApiUsage/Logging/LoggingExecutionContextScopeBuilderAnalyzer.cs

```diff
@@ -0,0 +1,147 @@
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.ValueTask" ),
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.ValueTask`1" )
+				)
+				.Where( x => x != null && x.Kind != SymbolKind.ErrorType )
+				.ToImmutableHashSet();
+
+			INamedTypeSymbol AsyncMethodBuilderAttribute = context.Compilation.GetTypeByMetadataName( "System.Runtime.CompilerServices.AsyncMethodBuilderAttribute" );
+
+			context.RegisterSyntaxNodeAction(
+				ctx => RunInvocationAnalysis(
+					context: ctx,
+					AsyncMethodBuilderAttribute: AsyncMethodBuilderAttribute,
+					ILoggingExecutionContextScopeBuilderRunSymbols: ILoggingExecutionContextScopeBuilderRunSymbols,
+					taskTypeBuiltins: taskTypeBuiltins,
+					invocationSyntax: ctx.Node as InvocationExpressionSyntax
+				),
+				SyntaxKind.InvocationExpression
+			);
+		}
+
+		private void RunInvocationAnalysis(
+			SyntaxNodeAnalysisContext context,
+			IImmutableSet<IMethodSymbol> ILoggingExecutionContextScopeBuilderRunSymbols,
+			INamedTypeSymbol AsyncMethodBuilderAttribute,
+			IImmutableSet<INamedTypeSymbol> taskTypeBuiltins,
+			InvocationExpressionSyntax invocationSyntax
+		) {
+			SemanticModel model = context.SemanticModel;
+
+			SymbolInfo methodSymbolInfo = model.GetSymbolInfo( invocationSyntax, context.CancellationToken );
+
+			bool isInvokingRun = ILoggingExecutionContextScopeBuilderRunSymbols.Contains( methodSymbolInfo.Symbol )
+				|| methodSymbolInfo
+					.CandidateSymbols
+					.Select( s => s.OriginalDefinition )
+					.Any( ILoggingExecutionContextScopeBuilderRunSymbols.Contains );
+			// Not Run or Run<T>
+			if( !isInvokingRun ) {
+				return;
+			}
+
+			var arguments = invocationSyntax.ArgumentList.Arguments;
+			if( arguments.Count != 1 ) {
+				return;
+			}
+
+			ArgumentSyntax actionArgument = arguments.First();
+			SymbolInfo argumentSymbolInfo = model.GetSymbolInfo(
+				actionArgument.Expression,
+				context.CancellationToken
+			);
+
+			ImmutableArray<IMethodSymbol> actionSymbolsToCheck;
+			if( argumentSymbolInfo.Symbol is IMethodSymbol argMethodSymbol ) {
+				actionSymbolsToCheck = ImmutableArray.Create( argMethodSymbol );
+			} else {
+				actionSymbolsToCheck = argumentSymbolInfo
+					.CandidateSymbols
+					.Where( s => s.Kind == SymbolKind.Method )
+					.Cast<IMethodSymbol>()
+					.ToImmutableArray();
+			}
+
+			foreach( IMethodSymbol actionSymbol in actionSymbolsToCheck ) {
+				ITypeSymbol actionReturnType = actionSymbol.ReturnType.OriginalDefinition;
+
+				if( taskTypeBuiltins.Contains( actionReturnType ) ) {
+					ReportDiagnostic(
+						context: context,
+						syntax: invocationSyntax
+					);
+					return;
+				}
+
+				// Check it it's a custom awaitable
+				IEnumerable<AttributeData> typeAttributes = actionReturnType.GetAttributes();
+				bool typeIsCustomAwaitable = typeAttributes.Any(
+					a => a.AttributeClass.OriginalDefinition == AsyncMethodBuilderAttribute
+				);
+				if( typeIsCustomAwaitable ) {

```

Comment:
I wasn't able to write a spec test that exercised this path, since it didn't want to resolve the type of the attribute in the spec file, regardless of whether I defined the namespace and type inline, or installed the nuget.